File: examples/python/examples/device_authorization.py
from __future__ import annotations

import argparse
import sys
from collections.abc import Callable
from pathlib import Path
from types import TracebackType

from qnbot_sdk import Sdk, SerialConnection, Side
from qnbot_sdk.glove import ActivationChallenge, GloveConfig


class _Cleanup:
    def __init__(self) -> None:
        self._primary: tuple[BaseException, TracebackType | None] | None = None
        self._cleanup_failure: RuntimeError | None = None

    def capture_primary(self, error: BaseException) -> None:
        if self._primary is None:
            self._primary = (error, error.__traceback__)

    def run(self, label: str, action: Callable[[], object]) -> bool:
        try:
            action()
            return True
        except BaseException as error:
            if self._cleanup_failure is None:
                self._cleanup_failure = RuntimeError(
                    f"qnbot example cleanup failed during {label} (details redacted)"
                )
            print(
                f"cleanup failed: {label} ({type(error).__name__})",
                file=sys.stderr,
            )
            return False

    def rethrow_if_failed(self) -> None:
        if self._primary is not None:
            error, traceback = self._primary
            raise error.with_traceback(traceback)
        if self._cleanup_failure is not None:
            raise self._cleanup_failure from None


def _cleanup_authorization_session(
    cleanup: _Cleanup,
    device: object,
    *,
    started: bool,
    factory_active: bool,
    connected: bool,
) -> tuple[bool, bool, bool]:
    if started and cleanup.run("device.stop()", device.stop):  # type: ignore[attr-defined]
        started = False
    if (
        not started
        and factory_active
        and cleanup.run(
            "device.exit_factory_calibration()",
            device.exit_factory_calibration,  # type: ignore[attr-defined]
        )
    ):
        factory_active = False
    if (
        not started
        and not factory_active
        and connected
        and cleanup.run(
            "device.disconnect()",
            device.disconnect,  # type: ignore[attr-defined]
        )
    ):
        connected = False
    return started, factory_active, connected


def create_sdk(port: str, side: Side) -> Sdk:
    return Sdk(
        devices=(
            GloveConfig(
                side=side,
                connection=SerialConnection(port=port),
            ),
        )
    )


def load_service_response(
    challenge: ActivationChallenge,
    operation: str,
    ticket_path: Path,
    signature_path: Path,
) -> tuple[bytes, bytes]:
    """Application integration point for service response files.

    Replace these file reads with the application's authenticated activation
    service client. Send the Challenge fields and operation to that service;
    Ticket and signature stay opaque to the SDK.
    """
    _ = challenge, operation
    return ticket_path.read_bytes(), signature_path.read_bytes()


def main() -> None:
    arguments = argparse.ArgumentParser(
        description="Authorize one serial glove with application-supplied service output"
    )
    arguments.add_argument("--port", required=True)
    arguments.add_argument(
        "--side",
        choices=(Side.LEFT.value, Side.RIGHT.value),
        required=True,
        help="Physical glove side",
    )
    arguments.add_argument(
        "--operation",
        choices=("permanent", "deactivate", "factory"),
        required=True,
    )
    arguments.add_argument("--ticket", type=Path, required=True)
    arguments.add_argument("--signature", type=Path, required=True)
    options = arguments.parse_args()

    sdk = create_sdk(options.port, Side(options.side))
    glove = sdk.glove()
    glove.connect()
    device = glove.device()
    connected = True
    started = False
    factory_active = False
    cleanup = _Cleanup()
    try:
        info = device.get_device_info()
        status = device.activation_status()
        print(
            f"device model={info.model} firmware={info.firmware_version} "
            f"activation={status.state.name}"
        )

        challenge = device.activation_challenge()
        ticket_bytes, signature = load_service_response(
            challenge,
            options.operation,
            options.ticket,
            options.signature,
        )
        if options.operation == "permanent":
            result = device.activate_permanently(ticket_bytes, signature)
            print(f"permanent activation={result.outcome.name}")
        elif options.operation == "deactivate":
            result = device.revoke_permanent_activation(ticket_bytes, signature)
            print(f"permanent deactivation={result.outcome.name}")
        else:
            factory = device.authorize_factory_calibration(ticket_bytes, signature)
            factory_active = factory.active
            print(f"factory authorization active={factory.active}")

        if options.operation != "deactivate":
            device.start()
            started = True
            print("telemetry started")
            device.stop()
            started = False
            print("telemetry stopped")

        if factory_active:
            device.exit_factory_calibration()
            factory_active = False
            print("factory authorization exited")

        device.disconnect()
        connected = False
    except BaseException as error:
        cleanup.capture_primary(error)

    started, factory_active, connected = _cleanup_authorization_session(
        cleanup,
        device,
        started=started,
        factory_active=factory_active,
        connected=connected,
    )
    if not connected:
        cleanup.run("glove.close()", glove.close)
    cleanup.rethrow_if_failed()


if __name__ == "__main__":
    main()

File: examples/cpp/examples/device_authorization.cpp
#include "device_authorization_cleanup.hpp"
#include "example_support.hpp"

#include <cstdlib>
#include <fstream>
#include <iostream>
#include <optional>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

namespace {

struct AuthorizationOptions {
    std::string port;
    std::optional<qnbot::Side> side;
    std::string operation;
    std::string ticket_path;
    std::string signature_path;
    bool validate_only = false;
};

AuthorizationOptions parse_options(int argc, char** argv) {
    AuthorizationOptions options;
    for (int index = 1; index < argc; ++index) {
        const std::string argument = argv[index];
        if (argument == "--port") {
            options.port = example::require_value(argc, argv, index, argument);
        } else if (argument == "--side") {
            options.side = example::parse_side(
                example::require_value(argc, argv, index, argument));
        } else if (argument == "--operation") {
            options.operation =
                example::require_value(argc, argv, index, argument);
        } else if (argument == "--ticket") {
            options.ticket_path =
                example::require_value(argc, argv, index, argument);
        } else if (argument == "--signature") {
            options.signature_path =
                example::require_value(argc, argv, index, argument);
        } else if (argument == "--validate") {
            options.validate_only = true;
        } else {
            throw std::invalid_argument("unknown argument: " + argument);
        }
    }
    if (options.port.empty()) throw std::invalid_argument("--port is required");
    if (!options.side) {
        throw std::invalid_argument("--side is required with --port");
    }
    if (options.operation != "permanent" && options.operation != "deactivate" &&
        options.operation != "factory") {
        throw std::invalid_argument(
            "--operation must be permanent, deactivate, or factory");
    }
    if (options.ticket_path.empty())
        throw std::invalid_argument("--ticket is required");
    if (options.signature_path.empty()) {
        throw std::invalid_argument("--signature is required");
    }
    return options;
}

std::vector<std::uint8_t> read_binary_file(const std::string& path) {
    std::ifstream input(path, std::ios::binary);
    if (!input)
        throw std::runtime_error("unable to open authorization response file");
    return {std::istreambuf_iterator<char>(input),
            std::istreambuf_iterator<char>()};
}

std::pair<std::vector<std::uint8_t>, std::vector<std::uint8_t>>
load_service_response(const qnbot::GloveActivationChallenge& challenge,
                      const AuthorizationOptions& options) {
    // Replace these file reads with the application's authenticated service
    // client. Pass the Challenge and operation; keep Ticket/signature opaque.
    static_cast<void>(challenge);
    return {read_binary_file(options.ticket_path),
            read_binary_file(options.signature_path)};
}

} // namespace

int main(int argc, char** argv) {
    try {
        const auto options = parse_options(argc, argv);
        qnbot::Sdk sdk(example::serial_config(options.port, *options.side));
        example::Cleanup cleanup;
        std::optional<qnbot::Glove> glove;
        std::optional<qnbot::GloveDevice> selected_device;
        bool connected = false;
        bool started = false;
        bool factory_active = false;
        try {
            glove.emplace(sdk.glove());
            if (!options.validate_only) {
                glove->connect();
                connected = true;
                selected_device.emplace(glove->device("primary"));
                auto& device = *selected_device;
                const auto info = device.get_device_info();
                const auto status = device.activation_status();
                std::cout << "device model=" << info.model
                          << " firmware=" << info.firmware_version
                          << " activated=" << status.activated << '\n';

                const auto challenge = device.activation_challenge();
                auto service_response =
                    load_service_response(challenge, options);
                if (options.operation == "permanent") {
                    const auto result = device.activate_permanently(
                        qnbot::ByteView(service_response.first),
                        qnbot::ByteView(service_response.second));
                    std::cout
                        << "permanent activated=" << result.status.activated
                        << '\n';
                } else if (options.operation == "deactivate") {
                    const auto result = device.revoke_permanent_activation(
                        qnbot::ByteView(service_response.first),
                        qnbot::ByteView(service_response.second));
                    std::cout
                        << "permanent deactivated=" << !result.status.activated
                        << '\n';
                } else {
                    const auto factory = device.authorize_factory_calibration(
                        qnbot::ByteView(service_response.first),
                        qnbot::ByteView(service_response.second));
                    factory_active = factory.active;
                    std::cout << "factory active=" << factory.active << '\n';
                }

                if (options.operation != "deactivate") {
                    device.start();
                    started = true;
                    std::cout << "telemetry started\n";
                    device.stop();
                    started = false;
                    std::cout << "telemetry stopped\n";
                }
                if (factory_active) {
                    static_cast<void>(device.exit_factory_calibration());
                    factory_active = false;
                }
                device.disconnect();
                connected = false;
            }
        } catch (...) {
            cleanup.capture_current("device authorization");
        }

        static_cast<void>(example::cleanup_authorization_session(
            cleanup, selected_device ? &*selected_device : nullptr,
            {started, factory_active, connected},
            [&] {
                if (glove) glove->close();
            },
            [&] { sdk.close(); }));
        cleanup.rethrow_if_failed();
        if (options.validate_only) {
            std::cout << "device authorization arguments valid\n";
        }
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        return example::report_error(error);
    }
}
